src/hotspot/share/runtime/os.cpp

```diff
@@ -714,30 +715,29 @@ void* os::realloc(void *memblock, size_t size, MEMFLAGS memflags, const NativeCa
       return NULL;
     }
 
-    const size_t old_size = MallocTracker::malloc_header(memblock)->size();
-
-    // De-account the old block from NMT *before* calling the real realloc(3) since it
-    // may invalidate old block including its header. This will also perform integrity checks
-    // on the old block (e.g. overwriters) and mark the old header as dead.
-    void* const old_outer_ptr = MemTracker::record_free(memblock);
+    MallocHeader* header = MallocTracker::malloc_header(memblock);
+    header->assert_block_integrity(); // Assert block hasn't been tampered with.
+    const FreePackage free_package = header->free_package();
+    header->mark_block_as_dead();
```

Comment:
Please revive the old comment, suitably adapted. It should still mostly apply, but please mention that de-accounting happens delayed since realloc may fail.